### PR TITLE
WIP: Oracle API indexing  implementation for ocean 

### PR DESCRIPTION
**Pull request by nagarajm22**
## Summary

- Oracle API indexing  implementation for ocean.

## RPCs

-  Oracle API indexing  implementation.

## Storage

- `rocksdb` indexing implementation for oracle api.

## Implications

- Storage
  - [ ] Database reindex required
  - [ ] Database reindex optional
  - [ ] Database reindex not required
  - [ ] None

- Consensus
  - [ ] Network upgrade required
  - [ ] Includes backward compatible changes
  - [ ] Includes consensus workarounds
  - [ ] Includes consensus refactors
  - [ ] None

### CODE REVIEW

**Review by canonbrother — APPROVED** [review]
utACK

**Comment by Jouzo** [comment]
Still need to sort cpp formatting. 
We also need API implementation for `/oracles` endpoint

File: lib/ain-ocean/src/indexer/oracle.rs [hunk 1]
```diff
@@ -39,11 +44,202 @@ impl Index for UpdateOracle {
...
+
+                    key: (
+                        token_price.token.clone(),
+                        token_amount.currency.clone(),
+                        set_data.oracle_id.to_string(),
+                    ),
+                    sort: hex::encode(ctx.block.height.to_string() + &ctx.tx.txid.to_string()),
+                    amount: token_amount.amount,
+                    currency: token_amount.currency.clone(),
+                    block: BlockContext {
```
Comment (Jouzo) [inline lib/ain-ocean/src/indexer/oracle.rs:227]: Can clone the `ctx.block` directly here

File: lib/ain-ocean/src/indexer/oracle.rs [hunk 2]
```diff
@@ -39,11 +44,202 @@ impl Index for UpdateOracle {
...
+                        hash: ctx.block.hash.clone(),
+                        height: ctx.block.height,
+                        median_time: ctx.block.median_time,
+                        time: ctx.block.time,
+                    },
+                    oracle_id: set_data.oracle_id.to_string(),
```
Comment (Jouzo) [inline lib/ain-ocean/src/indexer/oracle.rs:233]: All `oracle_id` can be kept as Txid

File: lib/ain-ocean/src/indexer/oracle.rs [hunk 3]
```diff
@@ -39,11 +44,202 @@ impl Index for UpdateOracle {
...
     }
 }
+
+fn map_price_aggregated(
+    services: &Arc<Services>,
+    ctx: &Context,
+    token: &str,
+    currency: &str,
+) -> Option<OraclePriceAggregated> {
+    // Convert Result to Option
```
Comment (Jouzo) [inline lib/ain-ocean/src/indexer/oracle.rs:89]: Let's not have it silently fail. Return a `Result<Option<OraclePriceAggregated>>` instead

File: lib/ain-ocean/src/indexer/oracle.rs [hunk 4]
```diff
@@ -39,11 +44,202 @@ impl Index for UpdateOracle {
 }
 
 impl Index for SetOracleData {
-    fn index(&self, services: &Arc<Services>, ctx: &Context) -> Result<()> {
-        todo!()
+    fn index(&self, services: &Arc<Services>, context: &Context) -> Result<()> {
+        let feeds = map_price_feeds(vec![self], vec![context])?;
```
Comment (Jouzo) [inline lib/ain-ocean/src/indexer/oracle.rs:48]: Why are we passing `vec![self]` ? What is this preparing for ?

File: lib/ain-ocean/src/indexer/oracle.rs [hunk 5]
```diff
@@ -39,11 +44,202 @@ impl Index for UpdateOracle {
...
+        let mut pairs: HashSet<(String, String)> = HashSet::new();
+        for feed in feeds {
+            pairs.insert((feed.token.clone(), feed.currency.clone()));
+            services.oracle_price_feed.by_id.put(&feed.id, &feed)?;
+        }
```
Comment (Jouzo) [inline lib/ain-ocean/src/indexer/oracle.rs:53]:
```suggestion
        let mut pairs = HashSet::new();
        for feed in feeds {
            services.oracle_price_feed.by_id.put(&feed.id, &feed)?;
            pairs.insert((feed.token, feed.currency));
        }
```

File: lib/ain-ocean/src/indexer/oracle.rs [hunk 6]
```diff
@@ -39,11 +44,202 @@ impl Index for UpdateOracle {
...
+
+    let result = oracle_token_currencies;
+    let mut aggregated = OraclePriceAggregatedAggregated {
+        amount: "0".to_string(),
+        weightage: 0,
+        oracles: OraclePriceAggregatedAggregatedOracles {
+            active: 0,
+            total: 0,
+        },
+    };
```
Comment (Jouzo) [inline lib/ain-ocean/src/indexer/oracle.rs:125]: Is this the expected values ? If so, we should implement Default directly

File: lib/ain-ocean/src/indexer/oracle.rs [hunk 7]
```diff
@@ -39,11 +44,202 @@ impl Index for UpdateOracle {
...
+                        .map(|data| Some(data))
+                },
+            ),
+            Err(err) => Err(err), // Pass along the error if there was one
+        };
+
+        let oracle_price_feed: Option<OraclePriceFeed> = match feeds {
+            Ok(Some(Some(feed))) => Some(feed),
+            _ => None,
+        };
```
Comment (Jouzo) [inline lib/ain-ocean/src/indexer/oracle.rs:152]: Same as above, shouldn't silence the error

**Pull request closed without merging**